maitake/src/wait/map.rs

```diff
@@ -0,0 +1,926 @@
+    Empty,
+
+    /// The Waiter has moved to the `WaitState::Waiting` state. We now
+    /// have the relevant waker, and are still waiting for data. This
+    /// corresponds to `WaitState::Waiting`.
+    Waiting(Waker),
+
+    /// The Waiter has received data, and is waiting for the woken task
+    /// to notice, and take the data by polling+completing the future.
+    /// This corresponds to `WaitState::Completed`.
+    DataReceived,
+
+    /// The waiter has received data, and already given it away, and has
+    /// no more data to give. This corresponds to `WaitState::Completed`.
+    Retreived,
+
+    /// The Queue the waiter is part of has been closed. No data will
+    /// be received from this future. This corresponds to
+    /// `WaitState::Completed`.
+    Closed,
+}
+
+// === impl WaitMap ===
+
+impl<K: PartialEq, V> WaitMap<K, V> {
+    /// Returns a new `WaitMap`.
+    #[must_use]
+    #[cfg(not(loom))]
+    pub const fn new() -> Self {
+        Self {
+            state: CachePadded::new(AtomicUsize::new(State::Empty.into_usize())),
+            queue: Mutex::new(List::new()),
+        }
+    }
+
+    /// Returns a new `WaitMap`.
+    #[must_use]
+    #[cfg(loom)]
+    pub fn new() -> Self {
+        Self {
+            state: CachePadded::new(AtomicUsize::new(State::Empty.into_usize())),
+            queue: Mutex::new(List::new()),
+        }
+    }
+
+    /// Wake a certain task in the queue.
+    ///
+    /// If the queue is empty, a wakeup is stored in the `WaitMap`, and the
+    /// next call to [`wait`] will complete immediately.
+    ///
+    /// [`wait`]: WaitMap::wait
+    #[inline]
+    pub fn wake(&self, key: &K, val: V) -> WakeOutcome<V> {
+        // snapshot the queue's current state.
+        let mut state = self.load();
+
+        // check if any tasks are currently waiting on this queue. if there are
+        // no waiting tasks, store the wakeup to be consumed by the next call to
+        // `wait`.
+        match state {
+            // Something is waiting!
+            State::Waiting => {}
+
+            // if the queue is closed, bail.
+            State::Closed => return WakeOutcome::Closed(val),
+
+            // if the queue is empty, bail.
+            State::Empty => return WakeOutcome::NoMatch(val),
+        }
+
+        // okay, there are tasks waiting on the queue; we must acquire the lock
+        // on the linked list and wake the next task from the queue.
+        let mut queue = self.queue.lock();
+        test_trace!("wake: -> locked");
+
+        // the queue's state may have changed while we were waiting to acquire
+        // the lock, so we need to acquire a new snapshot.
+        state = self.load();
+
+        if let Some(mut node) = self.node_match_locked(key, &mut *queue, state) {
+            let waker = Waiter::<K, V>::wake(node, &mut *queue, Wakeup::DataReceived);
+            unsafe {
+                node.as_mut().val.with_mut(|v| (*v).write(val));
```

Comment:
this is safe because we're holding the lock...maybe worth a comment?

Reply:
(if we stored the data in the `Node`, we could use `Waiter::with_node` to "prove" we're holding the lock...but maybe that's not worth it)